### DISCUSSION TITLE
how to manage session.begin() vs. autobegin

### DISCUSSION THREAD

**Discussion by robvdl** [sqlalchemy/sqlalchemy#6844] [Usage Questions]
### Describe the bug

I'm using SQLAlchemy 1.4 and connecting using future=True on the engine/session.

I am noticing that the session autobegin behaviour is triggered when doing simple fetch query and that is stating to get super annoying already.

I've got people telling me that other frameworks generally only do autobegin for updates. Doing an autobegin just for a fetch / select query seems excessive. And I am constantly running into the error "A transaction has already begun".

I'm actually struggling more with the new behaviour than the old because triggering this seems to be so easy if session autobegin kicks in the minute you do a select.

Is there no way to disable the autobegin behaviour because it's really annoying to be honest.

### To Reproduce

```python
engine = create_engine(connection_string, future=True)
session = Session(engine, future=True)
query = select(SomeModel).filter_by(...)

# why does this statement autobegin if it only does a select??
some_obj = session.execute(query).scalar_one_or_none()

# this will trigger a transaction has already begun.
with session.begin():
    pass
```

### Error

Constantly running into "A transaction has already begun" because simply querying triggers the autobegin behaviour.

### Versions

- OS: All
- Python: 3.7
- SQLAlchemy: 1.4
- Database: Postgres
- DBAPI:

**Comment by CaselIT** [comment]
Hi,

As mentioned in the change documentation, [here](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#change-5074) the session has always started a transaction by default. Nothing has changed in this regard, actually now the session creates a transaction on demand, instead of on creation.

What you are describing is the change that removed the usage of nested transactions, see [this entry in the docs](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#session-subtransactions). This [entry in the docs](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html?highlight=commit%20you%20go#session-level-vs-engine-level-transaction-control) may also be useful.

Regarding your question 
> why does this statement autobegin if it only does a select??

how does sqlalchemy know that the statement is only a select? It currently uses regexp to find out in autocommit mode (also deprecated), but that's fragile and can lead to hard to debug errors.

**Reply by robvdl** [reply]
Still doesn't really answer my question. I have a boss saying "DBI doesn't do that in Perl" so I'm a bit confused.

He thinks that "a session should be opened but you don't open a transaction until you do txn_do" in Perl DBI. I thought that this was analagous to "with session.begin()" as that automatically opens a transaction and I don't see "with session.transaction".

In SQLAlchemy I'm seeing a BEGIN statement the minute I do querying and then when I DO want to open a transaction it tells me off because it's already opened. Maybe he's wrong though.

I've worked around it for now, translating his old Perl code to Python, but I'm having to open transactions where he doesn't because of this autobegin behaviour.

**Reply by robvdl** [reply]
The question is how do I use the same session throughout but open the occasional transactions here and there.

Perhaps some docs on how to do that, only thing I can come up with is SAVEPOINT since nested transactions are gone, but I still kept running into "Transaction has already begun" everywhere even if I used session.begin(nested=True) more than once, it's super annoying the removal of nested transactions if you ask me.

**Reply by CaselIT** [reply]
> The question is how do I use the same session throughout but open the occasional transactions here and there.

With a session a transaction is always open. That's how a session has always behaved. The exception was the autocommit mode, but as mentioned, it's also deprecated. 

As described [here](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#session-subtransactions) and below by @lesnik512 you can use a context manager to handle the begin.
You may also want to commit/rollback the session before you need to open a new transaction block.

> it's super annoying the removal of nested transactions if you ask me.

The issue is that this patter does not do what most people thing it does. The following code
```py
session = Session()
some_obj = session.execute('select something).scalar_one_or_none()
with session.begin():  # this is nested trasnaction in the 1.3 mode
    session.execute('update some-table')
try:
  session.execute('select something-else').scalar_one()
except Exception:
  session.rollback()
```
will also rollback the `'update some-table'` statement, but that's not immediately clear by looking at the code, especially if that part is in a function somewhere. This is the main reason why it was removed.

**Reply by zzzeek** [reply]
> Still doesn't really answer my question. I have a boss saying "DBI doesn't do that in Perl" so I'm a bit confused.


this is because DBI does not default to non-autocommit the way [pep-249 does](https://www.python.org/dev/peps/pep-0249/#commit).    My work with SQLAlchemy started directly after I had worked with perl DBI for many years at MLB.com, one of the top 10 sites in the world at that time,  so this difference in behavior was in fact a huge paradigm shift for me as well, ultimately I like it **way** better but it took me years to come around and see this.   DBI actually works similarly to pep-249 when autocommit is off but they give you a "begin_work()" method that flips off autocommit mode for the scope of the next transaction so you sort of dont notice it.


> 
> He thinks that "a session should be opened but you don't open a transaction until you do txn_do" in Perl DBI. 

SQLAlchemy's "SessionTransaction" does not actually touch a database connection until needed.   `Session.begin()` does not procure any connectivity resources.  you'll note it's not awaitable in the asyncio api.


> In SQLAlchemy I'm seeing a BEGIN statement the minute I do querying and then when I DO want to open a transaction it tells me off because it's already opened. Maybe he's wrong though.

right it has "autobegin", which is a "new" feature only in that it does **not** begin the moment you create the `Session`, which is what it used to do.   that you can call `Session.begin()` at all when you arent in "autocommit" mode is in fact the *new* feature.     that said there's no need to call `Session.begin()` unless you are integrating into a framework of some kind that is looking for the "transaction start" event to occur at a specific point.
  

> 
> I've worked around it for now, translating his old Perl code to Python, but I'm having to open transactions where he doesn't because of this autobegin behaviour.

the DBI code used transactions also, it's just they were implicitly begun [**and** committed](https://metacpan.org/pod/DBI#AutoCommit) - note that this "commit" operation can be expensive on some backends compared to a rollback.  Notable quote from the linked page:

> Drivers should always default to AutoCommit mode (an unfortunate choice largely forced on the DBI by ODBC and JDBC conventions.)

I would link the above section to your boss so they can compare this to the decisions made by pep-249.

**Comment by lesnik512** [comment]
Hi, I use such context manager for transactions. It's async, but I think, it can be easily adopted to sync

```python
@asynccontextmanager
async def transaction(db: AsyncSession) -> AsyncGenerator[None, None]:
    """if select was called before than implicit transaction has already started"""
    if not db.in_transaction():
        async with db.begin():
            logger.debug("explicit transaction begin")
            yield
        logger.debug("explicit transaction commit")
    else:
        logger.debug("already in transaction")
        yield
        if db.in_transaction():
            await db.commit()
            logger.debug("implicit transaction commit")
```
And use it like this:
```python
    async with transaction(db):
        pass
```

**Reply by robvdl** [reply]
That's cool, we have a modern codebase, cli based, modern Python with lots of type hinting too, and single threaded single process which simplifies things. I can adapt this snippet myself thanks. It's a good starting point at least.

**Reply by Olegt0rr** [reply]
**async** def transaction :)

**Reply by CaselIT** [reply]
udated the example

**Comment by zzzeek** [comment]
two people asking basically the same question here so check out my answer at https://github.com/sqlalchemy/sqlalchemy/discussions/6864#discussioncomment-1142958

**Comment by tamanugi** [comment]
Hi  ✨ 

We are considering using `begin_nested ` to work around this problem.  
Is this solution correct? 

```python
engine = create_engine(connection_string, future=True)
session = Session(engine, future=True)
query = select(SomeModel).filter_by(...)

# why does this statement autobegin if it only does a select??
some_obj = session.execute(query).scalar_one_or_none()

# begin_nested not error 👍 
with session.begin_nested():
    pass
```

**Reply by CaselIT** [reply]
> We are considering using begin_nested to work around this problem.

I don't think it's a problem, that's just how the library and pep249 work by default (ie a transaction is always in progress)
Note that begin_nested creates a nested transaction on the database, so if that's what you want, sure use it

> why does this statement autobegin if it only does a select??

As repeated multiple times above, that's because sqlalchemy follows the pep249 that specifies that a transaction is always in progress.

**Reply by zzzeek** [reply]
> Hi sparkles
> 
> We are considering using `begin_nested ` to work around this problem. Is this solution correct?


if you are looking to avoid BEGIN for SELECT queries, your only option is to use [DBAPI level autocommit](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#setting-transaction-isolation-levels-dbapi-autocommit).  Otherwise the DBAPI emits BEGIN no matter what we do.   The linked section has detailed examples for how to set autocommit at every level (global, per session, per engine, per transaction, ~~per statement~~ (can't change autocommit mode per statement, whoops)) so should have you covered.

**Reply by tamanugi** [reply]
@CaselIT 
Thank you for quick reply ✨ 

> Note that begin_nested creates a nested transaction on the database

If I use `begin_nested` , what should I be aware of ?
I  notice that nested transactions are rolled back when the parent transaction is rolled back

So I am thinking that it might be better to use  https://github.com/sqlalchemy/sqlalchemy/discussions/6844#discussioncomment-1123727 rather than using begin_nested 🤔

**Reply by CaselIT** [reply]
> If I use begin_nested , what should I be aware of ?

The question is a bit broad... I just mentioned that begin nested has database impacts, that may or may not work in your use case, it's not just a sqlalchemy construct.

**Reply by tamanugi** [reply]
@CaselIT 
Thanks, You're right!
We would like to consider whether `nested_begin` is appropriate for our use case.
